Spigot-Server-Patches/0581-Improve-ServerGUI.patch

```diff
@@ -0,0 +1,453 @@
++    {
++        return ( ( tps > 21.0 ) ? "*" : "" ) + Math.min( Math.round( tps * 100.0 ) / 100.0, 20.0 ); // Paper - only print * at 21, we commonly peak to 20.02 as the tick sleep is not accurate enough, stop the noise
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/GuiStatsComponent.java b/src/main/java/net/minecraft/server/GuiStatsComponent.java
+index 859e31c63f94bdc7729c6d475990750b76e24b9c..0191358aab0dcaf30935c96ff7283fc2e4afd9ba 100644
+--- a/src/main/java/net/minecraft/server/GuiStatsComponent.java
++++ b/src/main/java/net/minecraft/server/GuiStatsComponent.java
+@@ -16,7 +16,7 @@ public class GuiStatsComponent extends JComponent {
+     });
+     private final int[] b = new int[256];
+     private int c;
+-    private final String[] d = new String[11];
++    private final String[] d = new String[12];
+     private final MinecraftServer e;
+     private final Timer f;
+ 
+@@ -35,8 +35,18 @@ public class GuiStatsComponent extends JComponent {
+     private void b() {
+         long i = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+ 
++        // Paper start - Add tps entry
++        double[] tps = org.bukkit.Bukkit.getTPS();
++        String[] tpsAvg = new String[tps.length];
++
++        for ( int g = 0; g < tps.length; g++) {
++            tpsAvg[g] = format( tps[g] );
++        }
+         this.d[0] = "Memory use: " + i / 1024L / 1024L + " mb (" + Runtime.getRuntime().freeMemory() * 100L / Runtime.getRuntime().maxMemory() + "% free)";
+         this.d[1] = "Avg tick: " + GuiStatsComponent.a.format(this.a(this.e.h) * 1.0E-6D) + " ms";
++        getStatEntries()[2] = "TPS from last 1m, 5m, 15m: " + String.join(", ", tpsAvg);
++        // Paper end
++
+         this.b[this.c++ & 255] = (int) (i * 100L / Runtime.getRuntime().maxMemory());
+         this.repaint();
+     }
+@@ -83,4 +93,10 @@ public class GuiStatsComponent extends JComponent {
+     public void a() {
+         this.f.stop();
+     }
++
++    public String[] getStatEntries() { return this.d; } // Paper - OBFHELPER
++    private static String format(double tps) // Paper - Add tps entry
++    {
++        return ( ( tps > 21.0 ) ? "*" : "" ) + Math.min( Math.round( tps * 100.0 ) / 100.0, 20.0 ); // Paper - only print * at 21, we commonly peak to 20.02 as the tick sleep is not accurate enough, stop the noise
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/ServerGUI.java b/src/main/java/net/minecraft/server/ServerGUI.java
+index e51434f2323bcb9385d11e99d0438d66434e3e63..032283c520536cafa075dfd3690590c03e488b94 100644
+--- a/src/main/java/net/minecraft/server/ServerGUI.java
++++ b/src/main/java/net/minecraft/server/ServerGUI.java
+@@ -2,15 +2,15 @@ package net.minecraft.server;
+ 
+ import com.google.common.collect.Lists;
+ import com.mojang.util.QueueLogAppender;
+-import java.awt.BorderLayout;
+-import java.awt.Component;
+-import java.awt.Dimension;
+-import java.awt.Font;
++
++import java.awt.*;
```

Comment:
No

Reply:
lmao sry again my ide is handling imports without me noticing.